### ISSUE TITLE
BatchNormalization Transform [feature request]

### ISSUE THREAD

**Issue by boathit** [pyro-ppl/pyro#1517]
It will be useful to add [BatchNorm Transform](https://www.tensorflow.org/api_docs/python/tf/contrib/distributions/bijectors/BatchNormalization) to stabilize the training in the normalizing flows, but I did not find it in the current dev branch, any plan to add the support?

**Comment by Kelvinson** [comment]
Hi, @eb8680 could you please any hints to  implement this feature? Add a module bijector like TFP or add it to the utility module? Thanks!

**Comment by fritzo** [comment]
This more naturally lives in [torch.distributions.transforms](https://pytorch.org/docs/stable/distributions.html#module-torch.distributions.transforms). Take a look a the [other transforms](https://github.com/pytorch/pytorch/blob/master/torch/distributions/transforms.py) to see how they're implemented.

**Comment by boathit** [comment]
All transforms implemented in `torch.distributions.tranforms` right now do not have the learnable parameters, this is not the case for BatchNorm Transform.

**Comment by fritzo** [comment]
@boathit oh I didn't realize there are trainable parameters. I think it's fine to add this to Pyro in a new file pyro/distributions/batch_norm_transform.py, and possibly move it upstream to torch.distributions once we work out the details.

I'm not sure the cleanest way to implement this, but it would be nice to leverage `AffineTransform` for `._inverse()` and `.log_abs_det_jacobian()`.

**Comment by boathit** [comment]
@fritzo I have managed to implement a simple [version](https://github.com/boathit/Deep-Probabilistic-Models/blob/master/maf/batchnorm.py) which assumes the input is just a D-dimension vector. I found it is able to significantly stabilize the training process in the [density estimation task](https://github.com/boathit/Deep-Probabilistic-Models/tree/master/maf).

Since the `Transform` is not a subclass of `nn.Module`, I dealt with this by first implementing a class `BatchNorm(nn.Module)` and then wrapping it with `BatchNormTransform(Transform)`.  In the training stage, I need to explicitly collect the parameters using `nn.ModuleList` in this [line](https://github.com/boathit/Deep-Probabilistic-Models/blob/master/maf/maskedAutoregressiveFlow.py#L14). I am not sure whether there is other more elegant way to handle this.

Anyway, I am looking forward a more general one in the next release of pyro.

**Comment by eb8680** [comment]
@boathit neat, thanks for sharing your code.  I don't think we'll get to this ourselves, but maybe you or @Kelvinson could turn your implementation into a PR?

**Comment by boathit** [comment]
@eb8680 Thanks. My only concern is that my current implementation is not general enough (assuming the input is a 1-D vector). I will try to figure out how to generalize it to handle arbitrary dimension tensor.

**Comment by fritzo** [comment]
@boathit if you'd be willing to submit a PR for your 1-D version (with tests :smile:) we could extend it to higher rank in a follow-up PR.

Also some comments:
- Can you help me understand why it needs to be an `nn.Module`? Do the beta and epsilon get trained simultaneously with other parameters?
- I think we you'd be able to use multiple inheritance if we had called `super(Transform, self).__init__()` in torch.distributions.transforms.Transform. This seems like a shortcoming in upstream code.

**Comment by boathit** [comment]
@fritzo I will be happy to make a PR.

> Can you help me understand why it needs to be an nn.Module? Do the beta and epsilon get trained simultaneously with other parameters?

Yes. gamma and beta are required to be trained simultaneously with all other parameters. It is easy to organize them by putting them into `nn.Module`. But of course I can move them into the `nn.Module` created in the `Transform`.

It would be desirable if `Transform` could inherit from `nn.Module`, then we can directly put all trainable parameters in the `Transform` without explicitly [creating a new Module](https://github.com/boathit/Deep-Probabilistic-Models/blob/master/maf/batchnorm.py#L50).

**Comment by stefanwebb** [comment]
@fritzo this was taken care of in a recent PR with the `TransformModule` class as you'll recall :)

Hi @boathit, I'm a contributor to Pyro who's been working on adding normalizing flow features. I don't mean to step on your turf, but I was thinking there might be a lower opportunity cost for me to add this feature to Pyro since I'm familiar with the code. Would you mind if I took your code and adapted it so it can be added to Pyro?

p.s. in [Line 201](https://github.com/boathit/Deep-Probabilistic-Models/blob/master/maf/normflow.py#L201) should it read `x.var` instead of `x.std`?

**Comment by boathit** [comment]
Hi Stefan, I will be very happy if you could add it to Pyro.

Stefan Webb <notifications@github.com> 于 2018年12月9日周日 12:37写道：

**Comment by boathit** [comment]
@stefanwebb You are correct, it should be `x.var` instead of `x.std`. The implementation also looks good to me. 

Besides, I find your IAF implementation may blow up the memory; as the cached `x` ([this line](https://github.com/uber/pyro/blob/dev/pyro/distributions/iaf.py#L92)) may never be popped up since we rarely do the inverse computation for the same `y` in the usage of IAF.

**Comment by stefanwebb** [comment]
This issue can be closed now after #1803
